### PR TITLE
Changes to getblocksubsidy RPC, and resolution of a problem with initializing chain params

### PR DESCRIPTION
**Pull request by daira**
This includes correct funding stream information in `getblocksubsidy` rpc calls,
and fixes an initialization order issue wherein key prefixes were not available
in time to allow for validation of funding stream addresses.

### CODE REVIEW

File: src/chainparams.cpp [hunk 1]
```diff
@@ -122,11 +122,27 @@ class CMainParams : public CChainParams {
...
         consensus.vUpgrades[Consensus::UPGRADE_CANOPY].nProtocolVersion = 170013;
-        consensus.vUpgrades[Consensus::UPGRADE_CANOPY].nActivationHeight =
-            Consensus::NetworkUpgrade::NO_ACTIVATION_HEIGHT;
+        consensus.vUpgrades[Consensus::UPGRADE_CANOPY].nActivationHeight = Consensus::NetworkUpgrade::NO_ACTIVATION_HEIGHT; //1046400;
 
         consensus.nFundingPeriodLength = consensus.nPostBlossomSubsidyHalvingInterval / 48;
 
+        // TODO: This `if` can be removed once canopy activation height is set.
+        if (consensus.vUpgrades[Consensus::UPGRADE_CANOPY].nActivationHeight != Consensus::NetworkUpgrade::NO_ACTIVATION_HEIGHT) {
+            std::vector<std::string> addresses(48, "t3Vz22vK5z2LcKEdg16Yv4FFneEL1zg9ojd");
```
Comment (nuttycom) [inline src/chainparams.cpp]: Is this intended to remain in the code, or just a placeholder for testing purposes?

**Review by nuttycom — COMMENTED** [review]
One minor question about the addresses added in this PR.

File: src/chainparams.cpp (hunk 1 above)
Comment (daira) [reply src/chainparams.cpp]: Just a placeholder.

**Review by daira — COMMENTED** [review]
ACK for @nuttycom's changes.

**Review by nuttycom — APPROVED** [review]

File: src/rpc/mining.cpp [hunk 2]
```diff
@@ -891,14 +893,27 @@ UniValue getblocksubsidy(const UniValue& params, bool fHelp)
...
-    result.pushKV("miner", ValueFromAmount(nReward));
+    if (canopyActive) {
+        UniValue fundingstreams(UniValue::VOBJ);
+        auto fsinfos = Consensus::GetActiveFundingStreams(nHeight, consensus);
+        for (auto fsinfo : fsinfos) {
+            CAmount nStreamAmount = fsinfo.Value(nBlockSubsidy);
+            fundingstreams.pushKV(fsinfo.recipient, ValueFromAmount(nStreamAmount));
+            nMinerReward -= nStreamAmount;
+        }
+        result.pushKV("fundingstreams", fundingstreams);
```
Comment (str4d) [inline src/rpc/mining.cpp]: This is currently structured as an object with one key per funding stream:
```
"fundingstreams": {
    "Recipient A": ZEC_VALUE_A,
    "Recipient B": ZEC_VALUE_B,
}
```
However, the recipient name is intended as an informational field, not a funding stream identifier; the latter would at a minimum have to mention ZIP 214.

I would prefer this to be structured as a list of objects:
```
"fundingstreams": [
    {
        "recipient": "Recipient A",
        "value": ZEC_VALUE_A
    },
    {
        "recipient": "Recipient B",
        "value": ZEC_VALUE_B
    }
]
```
This way, we don't need an explicit FS identifier, we don't have to worry about the recipient names changing, and we can extend this API with additional information later on.

**Review by str4d — CHANGES_REQUESTED** [review]
utACK after altering the `getblocksubsidy` output.

File: src/consensus/params.h [hunk 3]
```diff
@@ -123,6 +124,7 @@ class FundingStream
 
     static FundingStream ParseFundingStream(
         const Consensus::Params& params,
+        const KeyConstants& keyInfo,
```
Comment (therealyingtong) [inline src/consensus/params.h]:
```suggestion
        const KeyConstants& keyConstants,
```

File: src/consensus/params.h [hunk 4]
```diff
@@ -209,6 +211,7 @@ struct Params {
     int nFundingPeriodLength;
     boost::optional<FundingStream> vFundingStreams[MAX_FUNDING_STREAMS];
     void AddZIP207FundingStream(
+        const KeyConstants& keyInfo,
```
Comment (therealyingtong) [inline src/consensus/params.h]:
```suggestion
        const KeyConstants& keyConstants,
```

**Review by therealyingtong — APPROVED** [review]

File: src/rpc/mining.cpp (hunk 2 above)
Comment (daira) [reply src/rpc/mining.cpp]: Ok, will change. I'll also add a "specification" field since that's easy with this layout.

Comment (daira) [reply src/rpc/mining.cpp]:
```
$ src/zcash-cli getblocksubsidy 1046400
{
  "fundingstreams": [
    {
      "recipient": "Electric Coin Company",
      "specification": "https://zips.z.cash/zip-0214",
      "value": 0.21875000
    },
    {
      "recipient": "Zcash Foundation",
      "specification": "https://zips.z.cash/zip-0214",
      "value": 0.15625000
    },
    {
      "recipient": "Major Grants",
      "specification": "https://zips.z.cash/zip-0214",
      "value": 0.25000000
    }
  ],
  "miner": 2.50000000,
  "founders": 0.00000000
}
```

**Review by nuttycom — APPROVED** [review]
ACK changes post-review.

**Review by str4d — APPROVED** [review]
utACK

**Pull request merged**